LPJ-GUESS SDA Couplers

istfer opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- Please select appropriate Priority, Status,and Type labels-->
<!--- If you do not have permission to select labels please state which labels you would like -->

## Description
<!--- Describe your changes in detail -->

Here I am opening this PR as work in progress to publicly commit myself to it (some pun intended).

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Couplers for LPJ-GUESS to run PalEON-SDA.

## Review Time Estimate
<!---When do you want your code reviewed by?-->
- [ ] Immediately
- [ ] Within one week
- [ ] When possible
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue) <!-- please add issue number -->
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

istfer committed:
started new read state function

istfer committed:
add helper function that locates function

istfer committed:
add helper function that finds closing paranthesis

istfer committed:
add function that finds stream

istfer committed:
start reading

istfer committed:
first read

istfer committed:
extract typedef enums

istfer committed:
read typedef enum

istfer committed:
read Historic

istfer committed:
read vector

istfer committed:
read vector with const int

istfer committed:
read in parameters header too

istfer committed:
case fo seed

istfer committed:
ready to loop over stands

istfer committed:
start looping over patches

istfer committed:
reading individual class

istfer committed:
first pass at read individual

istfer committed:
finish individual

istfer committed:
few more flexibility

istfer committed:
read matrix

istfer committed:
getting ready to read sompool

istfer committed:
catch one more case

istfer committed:
finished soil

istfer committed:
adding more hardcoded constants for now

istfer committed:
getting ready to read fluxes

istfer committed:
finished patch

istfer committed:
finished reading all - first pass

MagicForrest committed:
Added allometry and allocation functions from LPJ-GUESS.  Also summing variables to a per-gridcells basis, and first start on updating the state.

istfer committed:
read meta.bin

MagicForrest committed:
allocation() function now correctly included

istfer committed:
k_latosa is sapwood_ratio in pecan

istfer committed:
Merge branch 'develop' of github.com:PecanProject/pecan into LPJGUESS_SDA

istfer committed:
small rearrangement for compilation

istfer committed:
add new par minmoist_est

MagicForrest committed:
Allocation of biomass now also works.  It seems to fail  when scaling biomass down ~0.55 or less. cmass_debt is not included in the calculation and scaled like the other pools.

istfer committed:
using phenology and leafphysiognomy

MagicForrest committed:
Added allometry() call.  Gives the target results for a biomass change doen to about 0.7.  Also updated the way of tracking individuals to better facilitate looping.

MagicForrest committed:
Factorised the adjustDensity and adjustBiomass code.

istfer committed:
few modifications for PalEON model version

istfer committed:
parsing files some more

MagicForrest committed:
Also scale and adjust N pools.

MagicForrest committed:
Mostly code tidying/documentation ahead of merging.

istfer committed:
Merge branch 'LPJGUESS_SDA' into LPJGUESS_SDA

istfer committed:
Merge pull request #2 from MagicForrest/LPJGUESS_SDA

First merge of LPJ-GUESS state updating code

istfer committed:
modifying function for PalEON version

istfer committed:
Merge branch 'LPJGUESS_SDA' of https://github.com/istfer/pecan into LPJGUESS_SDA

istfer committed:
read new PhotosynthesisResults

istfer committed:
updates for the new model version, first pass finished

MagicForrest committed:
Documented adjustBiomass() and cleaned the fucntion a little.

istfer committed:
start read_restart.LPJGUESS

istfer committed:
LPJGUESS uses exact days and will require split inputs function

istfer committed:
backbone of read_restart

MagicForrest committed:
Added HEIGHT_MAX as an argument to over-ride the LPJ-GUESS 150m max height.

MagicForrest committed:
Merge branch 'LPJGUESS_SDA' of github.com:MagicForrest/pecan into LPJGUESS_SDA

istfer committed:
start split inputs for LPJGUESS

istfer committed:
number of individuals are back

MagicForrest committed:
Just coding style

istfer committed:
Merge branch 'LPJGUESS_SDA' into LPJGUESS_SDA

istfer committed:
Merge pull request #3 from MagicForrest/LPJGUESS_SDA

merge hackathon pr

istfer committed:
define state year

istfer committed:
Merge branch 'LPJGUESS_SDA' of https://github.com/istfer/pecan into LPJGUESS_SDA

istfer committed:
add doc tags

istfer committed:
LPJGUESS package build

istfer committed:
no need for single quote

istfer committed:
Merge branch 'develop' of github.com:PecanProject/pecan into LPJGUESS_SDA

istfer committed:
Merge branch 'develop' of github.com:PecanProject/pecan into LPJGUESS_SDA

mdietze commented on models/lpjguess/R/read_state.R:
```diff
@@ -0,0 +1,1095 @@
+
+###################################### READ STATE
+
+library(stringr)
```
probably shouldn't be here

mdietze replied:
```suggestion
```

mdietze commented on models/lpjguess/R/split_inputs.LPJGUESS.R:
```diff
@@ -0,0 +1,131 @@
+## split LPJ-GUESS ncdf files into smaller time units to use in KF
```
FYI, this bit isn't needed unless the model, like sipnet, gets it start/end date info from its met files

infotroph committed:
Merge branch 'develop' into LPJGUESS_SDA

mdietze committed:
Merge branch 'develop' into LPJGUESS_SDA

mdietze committed:
Merge branch 'develop' into LPJGUESS_SDA

mdietze committed:
Update DESCRIPTION

Add "LinkingTo: Rcpp" to make usre Rcpp.h is found by compiler

yinghaoSunn committed:
Update pecan_package_dependencies.csv to pass the Github checks.

istfer committed:
Merge pull request #6 from yinghaoSunn/LPJGUESS_SDA

Update pecan_package_dependencies.csv to pass the Github checks.

github-actions[bot] committed:
define pi

yinghaoSunn committed:
Update RcppExports.cpp after running `make`

mdietze committed:
Update models/lpjguess/R/read_state.R

mdietze committed:
Merge branch 'develop' into LPJGUESS_SDA

yinghaoSunn committed:
Update NAMESPACE

Fix NAMESPACE imports for `str_match` and `glob2rx`

yinghaoSunn committed:
Update NAMESPACE “days_in_year”

Fix NAMESPACE imports for "days_in_year".
The "days_in_year" function was defined in "base/utils/R/days_in_year.R"

yinghaoSunn committed:
Update DESCRIPTION for stringr

yinghaoSunn committed:
Update NAMESPACE

yinghaoSunn committed:
Update NAMESPACE

It ‘s generated by roxygen2. I shouldn't edit it by hand so I change it back.

yinghaoSunn committed:
Fix documentation issues and update dependencies

- Added missing @description in update.state.LPJGUESS.R to fix roxygen2 errors.
- Included @name and @title for split_inputs.LPJGUESS.R to generate .Rd files correctly.
- Updated NAMESPACE to include proper imports for missing dependencies (PEcAn.utils, etc.).
- Resolved missing dependencies in DESCRIPTION (stringr).

istfer committed:
Merge pull request #7 from yinghaoSunn/LPJGUESS_SDA

Update RcppExports.cpp after running `make`

github-actions[bot] committed:
add dep

infotroph commented on models/lpjguess/R/adjust.biomass.LPJGUESS.R:
```diff
@@ -0,0 +1,145 @@
+# University of Illinois/NCSA Open Source License
+# which accompanies this distribution, and is available at
+# http://opensource.ncsa.illinois.edu/license.html
+#-------------------------------------------------------------------------------
```
License has changed to BSD3
```suggestion
```

infotroph commented on models/lpjguess/R/adjust.density.LPJGUESS.R:
```diff
@@ -0,0 +1,43 @@
+# University of Illinois/NCSA Open Source License
+# which accompanies this distribution, and is available at
+# http://opensource.ncsa.illinois.edu/license.html
+#-------------------------------------------------------------------------------
```
```suggestion
```

infotroph commented on models/lpjguess/R/update.state.LPJGUESS.R:
```diff
@@ -0,0 +1,302 @@
+# which accompanies this distribution, and is available at
+# http://opensource.ncsa.illinois.edu/license.html
+#-------------------------------------------------------------------------------
+
```
```suggestion
```

infotroph commented on models/lpjguess/R/calculateGridcellVariablePerPFT.LPJGUESS.R:
```diff
@@ -0,0 +1,136 @@
+# http://opensource.ncsa.illinois.edu/license.html
+#-------------------------------------------------------------------------------
+
+
```
```suggestion
```

infotroph reviewed (commented):
Removing a few copyright declarations that reference the previous NCSA license

infotroph commented on models/lpjguess/R/read_restart.LPJGUESS.R:
```diff
@@ -0,0 +1,61 @@
+      cmass_heart_perpft <- calculateGridcellVariablePerPFT(model.state = Gridcell_container, variable = "cmass_heart")
+      
+      cmass_wood <- cmass_sap_perpft + cmass_heart_perpft
+      cmass_wood <- udunits2::ud.convert(cmass_wood, "kg/m^2", "Mg/ha")
```
```suggestion
      cmass_wood <- PEcAn.utils::ud_convert(cmass_wood, "kg/m^2", "Mg/ha")
```

infotroph commented on models/lpjguess/DESCRIPTION:
```diff
@@ -14,7 +14,9 @@ Imports:
-    tibble
+    Rcpp (>= 1.0.1),
+    tibble,
+    stringr
```
Check results think `stringr` is unused

mdietze committed:
Update models/lpjguess/R/read_restart.LPJGUESS.R

Co-authored-by: Chris Black <chris@ckblack.org>

mdietze committed:
Update models/lpjguess/R/calculateGridcellVariablePerPFT.LPJGUESS.R

Co-authored-by: Chris Black <chris@ckblack.org>

mdietze committed:
Update models/lpjguess/R/adjust.biomass.LPJGUESS.R

Co-authored-by: Chris Black <chris@ckblack.org>

mdietze committed:
Update models/lpjguess/R/update.state.LPJGUESS.R

Co-authored-by: Chris Black <chris@ckblack.org>

mdietze committed:
Update models/lpjguess/R/adjust.density.LPJGUESS.R

Co-authored-by: Chris Black <chris@ckblack.org>

infotroph commented:
FYI when I try to build this locally on Mac using Clang 16.0.0, I'm getting a series of unknown type errors in allocation.LPJGUESS.cpp followed by "fatal error: too many errors emitted, stopping now [-ferror-limit=]". Happy to post more info / try to help debug if that's useful, but will wait until asked.

yinghaoSunn committed:
Fix: Address check_models issues
- Documented missing arguments for `update_state_LPJGUESS`, `write.config.LPJGUESS`, and `write.insfile.LPJGUESS`.
- Added missing dependencies to DESCRIPTION for `days_in_year`, `glob2rx`, and `str_match`.

yinghaoSunn committed:
1

yinghaoSunn committed:
Fix: Address check_models issues
- Renamed `update.state.LPJGUESS` to `update_state_LPJGUESS` to resolve S3 warnings.
- Added missing dependencies to NAMESPACE for `days_in_year`, `glob2rx`, and `str_match`.
- Added missing .Rd documentation  for functions in `read_state.R`.

yinghaoSunn committed:
Update `pecan_package_dependencies.csv` to pass the `test` check.

yinghaoSunn committed:
Add `@export` for `find_closing` to generate the function to NAMESPACE

yinghaoSunn committed:
Now `find_closing()` stops execution with an informative message if an invalid bracket is provided.

yinghaoSunn committed:
Now `find_closing()` stops execution with an informative message if an invalid bracket is provided.

yinghaoSunn committed:
Delete the example

yinghaoSunn committed:
Delete the examples.

yinghaoSunn committed:
Correct errors in Roxygen2 comments.

yinghaoSunn committed:
Delete extra roxygen2 comments.

Merge branch 'LPJGUESS_SDA' of https://github.com/istfer/pecan into LPJGUESS_SDA
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

istfer committed:
Merge pull request #8 from yinghaoSunn/LPJGUESS_SDA

Lpjguess sda
